### PR TITLE
Mid data

### PR DESCRIPTION
**Pull request by RobertoAldea**

### CODE REVIEW

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_common.py
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_common.py:1]: Add init functions to doc, specify if uses None or 0

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py [hunk 1]
```diff
@@ -110,15 +110,16 @@ def __init__(self) -> None:
         """Constructor of the class, the attributes will be create while running
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py]: It must use kargs

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py [hunk 2]
```diff
@@ -110,15 +110,16 @@ def __init__(self) -> None:
...
+class MidDataDeviceC: # pylint: disable=too-many-instance-attributes
     '''
     A class method that implements the MIDDataDevice class .
     '''
 
     # pylint: disable=too-many-arguments
-    def __init__(self, manufacturer : str| None= None, model :str| None= None,
-                serial_number : str| None= None, device_type : MidDataDeviceTypeE| None= None,
-                iface_name : str| None= None, mapping_names : Dict| None= None,
+    def __init__(self, dev_id: int|None = None, manufacturer : str| None= None,
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py:144]: In doc, specify that it uses None only on create function. Default param-> None

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 3]
```diff
@@ -137,11 +116,43 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
             curr_max (int): [description]
             curr_min (int): [description]
         '''
+        if ((volt_max is None and #pylint: disable=too-many-boolean-expressions
+             (volt_min is not None or curr_max is not None or curr_min is not None))
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Much easier, if any is none, then raise an error

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 4]
```diff
@@ -137,11 +116,43 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
...
+            or (volt_min is None and
+            (curr_max is not None or curr_min is not None or volt_max is not None))
+            or (curr_max is None and
+            (curr_min is not None or volt_max is not None or volt_min is not None))
+            or (curr_min is None and
+            (volt_max is not None or volt_min is not None or curr_max is not None))):
+            log.error("Invalid power range")
+            raise ValueError("Invalid power range")
+        if (volt_max is not None and volt_min is not None and volt_max < volt_min):
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Check only range is correct. Previously None type has been checked

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 5]
```diff
@@ -137,11 +116,43 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
...
+            log.error("Invalid voltage range")
+            raise ValueError("Invalid power range")
+        if (curr_max is not None and curr_min is not None and curr_max < curr_min):
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Same

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 6]
```diff
@@ -137,11 +116,43 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
...
+            bool: True if the current values are inside values of the other
+                instance, False otherwise
+        '''
+        res = False
+        if (self.volt_max <= aux_pwr_range.volt_max and self.volt_min >= aux_pwr_range.volt_min and
+            self.curr_max <= aux_pwr_range.curr_max and self.curr_min >= aux_pwr_range.curr_min):
+            res = True
+        return res
 
 class MidDataProfileC:
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py:240]: For instructions is not better to use and empty list instead of None

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py (hunk 2 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py:144]: Done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 3 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Makes sense, done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 4 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 5 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Idem

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 6 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py:240]: It actually doesn´t matter but could give a easier comprehension with less variable types. However, pylint gives a warning about dangerous default value

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py (hunk 1 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_devices.py]: Done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_common.py
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_common.py:1]: Done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 7]
```diff
@@ -137,11 +116,51 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
             curr_max (int): [description]
             curr_min (int): [description]
         '''
+        if ((volt_max is not None and volt_min is None) or #pylint: disable=too-many-boolean-expressions
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Move to fill functions

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 8]
```diff
@@ -137,11 +116,51 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
...
+
+        Args:
+            other (MidDataPwrRangeC): Instance to compare with
+
+        Returns:
+            bool: True if the current values are inside values of the other
+                instance, False otherwise
+        '''
+        res = False
+        if ((self.volt_max is None and self.curr_max is None) or
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Unnecesary if always use fill. Use property to read private attributes.

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 7 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Done

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py (hunk 8 above)
Comment (RobertoAldea) [reply code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Still needed if function is used before fill.
Properties added

File: code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py [hunk 9]
```diff
@@ -116,22 +116,86 @@ def __init__(self, volt_max : int|None = None, volt_min : int|None = None,
             curr_max (int): [description]
             curr_min (int): [description]
         '''
+        self.__volt_max : int|None = volt_max
+        self.__volt_min : int|None = volt_min
+        self.__curr_max : int|None = curr_max
+        self.__curr_min : int|None = curr_min
```
Comment (mariuscrsn) [inline code/src/wattrex_battery_cycler/mid/mid_data/mid_data_experiment.py]: Call fill_voltage and fill_current on init

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

**Review by mariuscrsn — APPROVED** [review]

**Review by mariuscrsn — APPROVED** [review]

**Pull request merged**